feat(twilight-model)!: Implement additional select menu types

archer-321 opened the pull request:
This merge request implements Discord's new select menu types (user, role, mentionable, and channel) using the fully-flattened approach outlined in [this RFC](https://github.com/twilight-rs/twilight/discussions/2217).

# General idea

The general idea is to implement all variant-specific fields as `Option`s inside the `SelectMenu` struct. While this might require runtime validation, it's more in line with the rest of the Twilight ecosystem. Moreover, it doesn't duplicate fields that an approach with separate structs would duplicate.

Notably, this differs from the RFC mentioned above. This RFC was indirectly denied with [this review](https://github.com/twilight-rs/twilight/pull/2219#pullrequestreview-1480445162).

# Breaking changes

This PR introduces multiple breaking changes:

- `ComponentType`'s `SelectMenu` variant was renamed, and the enum was extended (breaking as it's not `#[non_exhaustive]`).
- `SelectMenu`'s `options` field was wrapped in an `Option`.
- New `channel_types` and `kind` fields were introduced in `SelectMenu`.
- `twilight_model::application::interaction::application_command::{CommandInteractionDataResolved, InteractionChannel, InteractionMember}` were moved into `twilight_model::application::interaction`.
- `CommandInteractionDataResolved` was renamed to `InteractionDataResolved`.
- `MessageComponentInteractionData` no longer implements `Eq + Hash`.
- A new `resolved` field was introduced in `MessageComponent`.
- `InteractionData::MessageComponent`'s single field was `Box`-ed.

If I missed any breaking changes, please point them out.

# Related work

- This PR closes #2047.
- As mentioned above, this implements an idea outlined in #2217.
- The same feature was previously implemented in #2051, but that PR was seemingly abandoned.
- This feature is defined by [Discord's documentation](https://discord.com/developers/docs/interactions/message-components#select-menus)
- I created a throwaway bot to test this PR here: <https://github.com/archer-321/twilight-select-menus>.

# Tests

This PR only updates the existing tests. If you'd like me to add other tests to verify the behaviour introduced by this patch, please feel free to request changes.

# Signed commits

I'm unsure whether this is possible with Twilight's PR flow, but please ask me to squash any commits or edit commit messages if necessary. I prefer all commits attributed to me to be signed with my GPG key :nerd_face: 

# Labels

Based on the current state, I believe you should assign the `d-api` and `d-breaking` labels.

baptiste0928 reviewed (commented):
Thanks for this PR, it looks good to me overall!

It seems that you should also add a `resolved` field to `MessageComponentInteractionData` ([Discord documentation](https://discord.com/developers/docs/interactions/message-components#select-menu-object-select-menu-resolved-object)). This has been implemented in #2051 if you need a reference.

> Moreover, those instances are Box-ed inside the enum to avoid unnecessarily increasing the enum's memory footprint if Discord adds new fields in the future.

This need to be discussed, but I personally think it just adds unnecessary complexity. Discord is unlikely to add fields that make the enum very large - and we can't anticipate everything anyway.

archer-321 commented on twilight-model/src/channel/message/component/mod.rs:
```diff
@@ -436,28 +474,47 @@ impl<'de> Visitor<'de> for ComponentVisitor {
+                    }
+                    // We'll only take the branch below if we added a type above and forgot to implement it here. I.e.,
+                    // we should never end up here.
+                    _ => panic!("missing select menu implementation"),
```
Thanks for the suggestion; I snuck it into b545e9e :+1:

archer-321 commented:
> It seems that you should also add a `resolved` field to `MessageComponentInteractionData` ([Discord documentation](https://discord.com/developers/docs/interactions/message-components#select-menu-object-select-menu-resolved-object)). This has been implemented in #2051 if you need a reference.

Indeed! I'm not sure how I missed this. I took the opportunity to create a small throwaway bot here: <https://github.com/archer-321/twilight-select-menus>.

It responds to simple legacy text commands and sends a message with all select menu types.
<details>
<summary>Screenshot</summary>

![A screenshot showing the bot's message with five select menus](https://i.imgur.com/dFqLzWg.png)

</details>

> This need to be discussed, but I personally think it just adds unnecessary complexity. Discord is unlikely to add fields that make the enum very large - and we can't anticipate everything anyway.

I agree that it adds extra complexity and might be unnecessary. Still, I'm unsure whether the decreased complexity of removing it outweighs the benefit of having each variant be the size of a pointer. Let's wait for other reviews c:

baptiste0928 commented:
> Indeed! I'm not sure how I missed this. I took the opportunity to create a small throwaway bot here: https://github.com/archer-321/twilight-select-menus.

It's not documented in [Message Component Data Structure](https://discord.com/developers/docs/interactions/receiving-and-responding#interaction-object-message-component-data-structure), I only saw it because the previous PR had implemented it :sweat_smile: 

Thanks for the throwaway bot, it's really useful for reviewing :)

> the benefit of having each variant be the size of a pointer.

The enum is 32 bytes without `Box`, 16 bytes with. There is no benefit, it's even worse because you have to make a heap allocation.

HTG-YT reviewed (approved):
LGTM.

Gelbpunkt reviewed (changes requested):
I fundamentally disagree with the design choices in this PR, for several reasons. By "the design choices", I mean not flattening the struct.

1. If we were to make the change from flattened structs *back* to enums, that should be done for all model structs and not for a single struct. That is inconsistent and, in my view, a code smell.
2. We have in the past switched from enums to flattened structs for things like channel due to "[flattening] will prevent breaking changes and API oddities in the future" (a0ec10969bf2db4f624e7f10bb809f4d11545058). Flip-flopping back and forth is not really desirable for maintainers *or* users.
3. Field access gets really awkward with inner enums for fields present conditionally. `some_select_menu.options` as `Option<Vec<SelectMenuOption>>` makes accessing the field much easier than `if let` matching on parts of a struct which are contained in an enum.
4. Flattened structs model the Discord API closer and will be less likely to cause breaking changes (this kind of repeats the argument from 2.). If `options` is ever going to be present for other types than 3, it won't require breaking public API changes. This also models the Discord documentation a bit closer ([link](https://discord.com/developers/docs/interactions/message-components#select-menu-object-select-menu-structure)).

TLDR: I don't see the added value at all. I'd be happy to accept this if we would just include the struct fields flattened as options, or if we revisit the entire model crate and make serious future commitments instead of changing model philosophy every few releases.

HTG-YT commented:
> I fundamentally disagree with the design choices in this PR, for several reasons. By "the design choices", I mean not flattening the struct.
> 
> 1. If we were to make the change from flattened structs _back_ to enums, that should be done for all model structs and not for a single struct. That is inconsistent and, in my view, a code smell.
> 2. We have in the past switched from enums to flattened structs for things like channel due to "[flattening] will prevent breaking changes and API oddities in the future" ([a0ec109](https://github.com/twilight-rs/twilight/commit/a0ec10969bf2db4f624e7f10bb809f4d11545058)). Flip-flopping back and forth is not really desirable for maintainers _or_ users.
> 3. Field access gets really awkward with inner enums for fields present conditionally. `some_select_menu.options` as `Option<Vec<SelectMenuOption>>` makes accessing the field much easier than `if let` matching on parts of a struct which are contained in an enum.
> 4. Flattened structs model the Discord API closer and will be less likely to cause breaking changes (this kind of repeats the argument from 2.). If `options` is ever going to be present for other types than 3, it won't require breaking public API changes. This also models the Discord documentation a bit closer ([link](https://discord.com/developers/docs/interactions/message-components#select-menu-object-select-menu-structure)).
> 
> TLDR: I don't see the added value at all. I'd be happy to accept this if we would just include the struct fields flattened as options, or if we revisit the entire model crate and make serious future commitments instead of changing model philosophy every few releases.

To be fair. I would agree if the models involved in the enum have more fields in common - but this does not seem to be the case. I believe that if there are many common fields for different types in the enum, then sure flattening would be better (reduce duplication and also memory footprint). But in this case I do not see much point flattening albeit conflicting with the current design choices - as far as I know there are not much fields that are in common between different select menu types (perhaps only 1 field?).

Not really sure to be honest, but this indeed can spark further discussion.

Gelbpunkt commented:
There are 6 common fields and 2 conditional ones, you can check it [here](https://discord.com/developers/docs/interactions/message-components#select-menu-object-select-menu-structure).

ImDarkDiamond commented:
I'm also for flattening. After using the flattened channel struct for a while now I find it much nicer than matching on an enum

HTG-YT reviewed (commented):
As per comments made by @Gelbpunkt, flattening the structures will align better with the Discord API model as well as our design philosophy for model structures.

archer-321 commented:
> TLDR: I don't see the added value at all

I'm a huge fan of static enforcement for types, and I aim to implement it through my libraries wherever possible - even if it comes at the cost of increasing the required boilerplate. E.g. it's less convenient to match an `Option` than being able to dereference a pointer, but `Option`s prevent `null` pointer dereferencing at compile time.

That said, it's way less dramatic with the entirely flattened approach. Instead of crashing the program, the struct would allow users to specify invalid select menus (e.g. a channel select menu that tries to set the `options` field).
Considering that Discord would return an error if a developer tries to do this, I can see how the increased convenience can outweigh the added value of static enforcement.

Of course, I can flatten the struct in this PR, but I have a couple of questions regarding the desired implementation:

- How would users specify which select menu type they want to implement? User, role, and mentionable select menus don't add any type-specific fields, so there needs to be an additional field (e.g. `kind`) to select the type. However, having another `kind` type would duplicate the `kind` already present in the component (de-)serialization. Considering that Discord assigns different `type` values to the different select menu types, this approach doesn't follow the upstream API either.
- As outlined in https://github.com/twilight-rs/twilight/pull/2051#pullrequestreview-1251821104, "if we leave options unchanged, then if any other select menu type supports options, the type of option (string select menu options) won't fit". Should I follow the approach explained in this comment to move the implementation of the current options into a separate enum?

laralove143 commented:
i think this discussion should be moved to #2217 and that discussion should be renamed to reflect an overall consistent change in the model crate and then we should merge or request changes in the pr before merging it based on the result

we should first decide how we handle rfcs though

HTG-YT commented:
> i think this discussion should be moved to #2217 and that discussion should be renamed to reflect an overall consistent change in the model crate and then we should merge or request changes in the pr before merging it based on the result
> 
> we should first decide how we handle rfcs though

Do we want a similar procedure with that of Rust’s?

laralove143 commented:
> > i think this discussion should be moved to #2217 and that discussion should be renamed to reflect an overall consistent change in the model crate and then we should merge or request changes in the pr before merging it based on the result
> 
> > 
> 
> > we should first decide how we handle rfcs though
> 
> 
> 
> Do we want a similar procedure with that of Rust’s?

partially yeah i agree but i think we can keep it simpler since i scope is smaller i think the flow should be like

rfc -> contributing.md (if accepted) -> adapting twilight for rfc changes

archer-321 commented:
Could we progress with the RFC or this PR? If the static enforcement introduced by this change isn't an option because it would differ from the rest of the library, I'd be OK with implementing the entirely flattened approach. However, I'm still unsure about the two questions listed in <https://github.com/twilight-rs/twilight/pull/2219#issuecomment-1592181382> :worried:

In my opinion, an RFC wouldn't be needed when going for the flattened approach that aligns with other types in Twilight. Instead, a more general RFC should discuss the benefits and drawbacks of changing structs to a partially flattened implementation throughout the library, in this case.

HTG-YT commented:
> Could we progress with the RFC or this PR? If the static enforcement introduced by this change isn't an option because it would differ from the rest of the library, I'd be OK with implementing the entirely flattened approach. However, I'm still unsure about the two questions listed in [#2219 (comment)](https://github.com/twilight-rs/twilight/pull/2219#issuecomment-1592181382) 😟
> 
> In my opinion, an RFC wouldn't be needed when going for the flattened approach that aligns with other types in Twilight. Instead, a more general RFC should discuss the benefits and drawbacks of changing structs to a partially flattened implementation throughout the library, in this case.

Yeah, it would be nice if you could progress with implementing the flattened approach.

archer-321 commented:
I'm sorry if my last message sounded impatient or demanding; that wasn't my intention.

I agree; for now, an approach that's more in line with the rest of the library is probably best. However, I still have the two questions I linked to in my last message.

Specifically, I still have the two questions quoted below :arrow_down:

> - How would users specify which select menu type they want to implement? User, role, and mentionable select menus don't add any type-specific fields, so there needs to be an additional field (e.g. kind) to select the type. However, having another kind type would duplicate the kind already present in the component (de-)serialization. Considering that Discord assigns different type values to the different select menu types, this approach doesn't follow the upstream API either.
> - As outlined in
https://github.com/twilight-rs/twilight/pull/2051#pullrequestreview-1251821104, "if we leave options unchanged, then if any other select menu type supports options, the type of option (string select menu options) won't fit". Should I follow the approach explained in this comment to move the implementation of the current options into a separate enum?

I can't think of a satisfying implementation that addresses both problems (preventing a duplicate `kind` field and future-proofing the API if Discord adds more select menu options).

Before I spend more hours creating another patch that fundamentally doesn't align with the crate's vision, I'd like some input before I start changing the API again. I hope you understand :slightly_smiling_face:

baptiste0928 commented:
> How would users specify which select menu type they want to implement? 

A `kind` field seems like a good solution to me. There is no duplication in the public API (except for the `kind()` method), only in the implementation, so it's not confusing imo.

> Should I follow the approach explained in this comment to move the implementation of the current options into a separate enum?

Personally, I think it's best to focus on a simple implementation. We can't anticipate all the changes Discord will make, so it's not worth making our API more complicated for changes we don't know will happen. I'd be curious to know what others think about this.

archer-321 commented:
I implemented the entirely flattened approach requested above. If I missed something, please let me know.

> A `kind` field seems like a good solution to me.

Done :heavy_check_mark: 

> Personally, I think it's best to focus on a simple implementation. We can't anticipate all the changes Discord will make, so it's not worth making our API more complicated for changes we don't know will happen. I'd be curious to know what others think about this.

Considering that the API isn't stable across releases at this point in Twilight's development, this would likely be overkill. For now, I implemented the fields directly.

Gelbpunkt commented on twilight-model/src/channel/message/component/select_menu.rs:
```diff
@@ -1,27 +1,59 @@
     pub disabled: bool,
+    /// This select menu's type.
+    ///
+    /// [Text select menus](SelectMenuType::Text) *must* also set the `options` field.
```
I'm not a particular fan of the language here and/or mentioning this in this place at all but I'll leave that for others to decide

archer-321 replied:
Considering it's already mentioned in the `option` field's documentation, I agree this is unnecessary.

I removed it in ab68bfb6.

Gelbpunkt commented on twilight-model/src/channel/message/component/select_menu.rs:
```diff
@@ -1,27 +1,59 @@
 pub struct SelectMenu {
+    /// An optional list of channel types.
+    ///
+    /// This option is only used for [channel select menus](SelectMenuType::Channel).
```
Precedent language from [Channel](https://github.com/twilight-rs/twilight/blob/main/twilight-model/src/channel/mod.rs#L117) would be "This is only applicable to channel select menus", but this sounds okay too

archer-321 replied:
Changed :white_check_mark:

Gelbpunkt commented on twilight-model/src/channel/message/component/select_menu.rs:
```diff
@@ -1,27 +1,59 @@
-    pub options: Vec<SelectMenuOption>,
+    /// A list of available options.
+    ///
+    /// This value is only used and required by [text select menus](SelectMenuType::Text).
```
Earlier you use "option" and "field", here you use "value". Could we unify the language?

archer-321 replied:
I removed `value`, `field`, and `option` entirely from the documentation comments. Do you think this would address this concern?

Gelbpunkt commented on twilight-model/src/channel/message/component/select_menu.rs:
```diff
@@ -1,27 +1,59 @@
+pub enum SelectMenuType {
+    /// Select menus with a text-based `options` list.
+    ///
+    /// Select menus of this `kind` *must* set the `options` field to specify the options users
```
I don't think we should be writing API documentation for Discord in all places imaginable, one should be enough on the `options` field

archer-321 replied:
I mention this in the enum documentation, as users may only read its documentation to see which select menu types are available.
Now that the `options` field is an `Option<_>`, I think the duplicated documentation is worth it if it saves some people from debugging a Discord API error.
However, I'm open to talking about this instance.

Gelbpunkt replied:
any other opinions on this or do we leave it as is?

HTG-YT replied:
I agree upon that this duplicated documentation is worth it; in terms of helping users understand the requirements on fields for specific select menu types.

As such, I believe that we can keep this as is.

Gelbpunkt commented on twilight-model/src/channel/message/component/select_menu.rs:
```diff
@@ -1,27 +1,59 @@
+    Mentionable,
+    /// Channel select menus.
+    ///
+    /// Select menus of this `kind` *can* use the `channel_types` field to specify which types of
```
Same here, this is the documentation for the `type` enum, not the struct field

archer-321 replied:
I agree; I removed this line.

Gelbpunkt reviewed (approved):
Overall code LGTM, I have some nits on documentation but I'm not a native speaker and there are other people who might disagree so I'd be fine with merging this as-is, consider my comments as what went through my mind when reading this

vilgotf commented:
This PR has three approvals but is still labled as unapproved? What's holding this up?

Gelbpunkt commented:
> This PR has three approvals but is still labled as unapproved. What's holding this up?

Nothing, I forgot to remove the label after the PR switched to flattened models

HTG-YT commented:
> > This PR has three approvals but is still labled as unapproved. What's holding this up?
> 
> Nothing, I forgot to remove the label after the PR switched to flattened models

Then I presume that this pull request can be merged.

vilgotf commented:
Amazing work, thank you so much! 🎉
